Phase transition test

lhy11009 opened the pull request:
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

*Describe what you did in this PR and why you did it.*

### Before your first pull request:

* [x] I have read the guidelines in our [CONTRIBUTING.md](../CONTRIBUTING.md) document.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../doc/modules/changes) directory that will inform other users of my change.

gassmoeller commented:
/rebuild

lhy11009 commented on source/material_model/rheology/diffusion_creep.cc:
```diff
@@ -36,24 +37,44 @@ namespace aspect
-                                              const unsigned int composition) const
-      {
+                                              const unsigned int composition,
+                                              const std::pair<std::vector<double>*, const std::vector<unsigned int>*> *gamma_inputs) const
```
Hi Rene

I am working on the pull request row, but I find out that in gamma_inputs with the form of  'const std::pair<std::vector<double>*, const std::vector<unsigned int>*> *', the two vectors are not the same size. That is because the first one has the length of phases while the second one has the length of compositions. Thus, it can not be changed into the other form of it. Please let me know what you think of it.

gassmoeller replied:
Hi Haoyuan, sorry for coming back late to your comment. If the two vectors have different lengths and content that is not directly associated with each other that suggest that they maybe should not be a combined parameter, but instead two separate ones (one called `n_phases_per_composition` and one `phase_function_values`), do you think that makes sense?

lhy11009 commented:
Hi Rene, I have updated the pull request, I still use the same data type for phase_inputs but I get rid of all the pointers and use references instead. Menno told me about a way to change that to vector <pair<double, unsigned int>> and it is helpful. However, that may require we also change how we read the input and calculate the phase function value. We can discuss this issue.

Have a nice day.

lhy11009 commented:
Hi, Rene, I put a change_log in the directory, also check if that is okay for me, thanks
Your suggestion is great if we could pass these two vectors separately. But as we need to maintain the interface of some functions by giving default value, I think it doesn't make much sense to have two of them, because, inside the function, we could tell what happens by looking at just one vector and see if it has its default value. So I would still maintain the pair<vector<double>, const vecotr<unsigned int> > structure. We can discuss more if you have time.

lhy11009 commented:
@gassmoeller 
Hi Rene

It seems like there are some tests failed. I will check on that. And my question is how could I run all
the tests on my computer to check beforehand next I upload a request?

lhy11009 commented:
@gassmoeller 

I looked at the user manual, it says I could do

crest 

under my build directory. And

make setup_tests

to run all tests.
However, this doesn't seem to be the case because the time to run is too instant to be an entire test suite.
Moreover, how could I run a single test if I want to check the failed tests? I tried to directly run a single test in the tests folder, but I don't know how to compile those .cc files. The CMakeLists.txt file there gives me a Makefile that seems to do nothing when. Should I modify some other CMakeLists.tx files like those I can find in cookbooks? That could be done but doesn't seem like one of the meant-to-be-done things. Please tell me what you think, thanks

gassmoeller commented:
`ctest` by default will only run a small amount of tests (one), but after you run `make setup_tests` running `ctest` again should run all tests. Do you get something like the following output when running `make setup_tests`?
```
[100%] Enabling all tests ...
     Found 766 tests.
[100%] Built target setup_tests
```

Running specific tests works with `ctest -R test_name` as explained in the manual: http://www.math.clemson.edu/~heister/manual.pdf#sec%3Arunning_tests

The .cc files in the test folder are compiled by ctest and will be placed in a subdirectory `tests` of the directory where you compiled aspect. There will also be a file called `test_name.x.prm` that you can use to run the specific tests outside of ctest.

Your specific test results will vary from our reference tester, dont worry about that, but your new tests crash and you should check why that is by running them on your machine (test results: https://jenkins.geodynamics.org/blue/organizations/jenkins/aspect/detail/PR-3313/44/tests).

lhy11009 commented:
I captured my output:

make setup_test

---------------------------------------------------------
Could not read from count file.
Enabling all tests ...
     Found 765 tests.
Built target setup_tests
---------------------------------------------------------

ctest

---------------------------------------------------------
Test project /home/lochy/softwares/aspect/build
    Start 1: unit_tests
1/2 Test #1: unit_tests .......................   Passed    0.22 sec
    Start 2: quick_mpi
2/2 Test #2: quick_mpi ........................   Passed    0.37 sec

100% tests passed, 0 tests failed out of 2

Total Test time (real) =   0.60 sec
----------------------------------------------------------



and then:

ls tests

---------------------------------------------------------
aspect
CMakeCache.txt
CMakeFiles
cmake_install.cmake
CTestTestfile.cmake
libquick_mpi.so
Makefile
output-quick_mpi
quick_mpi.x.prm
setup_tests.log
---------------------------------------------------------

So there is only one "quick_mpi.x.prm". I tried to change to another branch but the only difference is on the number "765" which becomes "766". Is there something wrong with "Could not read from count file."? How to fix it?

gassmoeller commented:
Did you run cmake or ctest inside the `tests/` directory at some point? This might confuse the process. If so you can do the following to fix it: `rm -rf tests; git checkout tests`

lhy11009 commented:
@gassmoeller 
Thanks, Rene. You are right. I should have taken a careful look at the user manual before making some damn moves. -.-||

lhy11009 commented:
@gassmoeller 
Hi Rene

I found that a couple of tests starting with "visco_plastic_phases"  in their name haven't included the inputs for rheology and they keep giving errors.

And, in the current implementation when phases are included, it turns out that we need to give all the variables for flow law, both diffusion and dislocation creep even if a single dislocation creep is applied. I think we need to change that because that's redundant, do you agree?

What do you say, one way is to fix all the texts to fit the current implementation, the other is to fix the inputs, to improve it so that it could be more flexible. I believe I could do either of them,  but which is better?

lhy11009 commented:
@gassmoeller 
I think it would be better to make the input method more flexible to take only one value, so I have added a new commit. This also fixed the problem where the whole information of flow law must be included in the input file.

lhy11009 commented:
@gassmoeller 
Now I could get the right result from the tests I run.
However, this "visco_plastic_phases" test failed while another "visco_plastic_phases_compositions" test passed. When I compare the outputs of those two, I couldn't tell why the first one failed while the second one passed. Can you take a quick look? 
[output-visco_plastic_phases.zip](https://github.com/geodynamics/aspect/files/4384639/output-visco_plastic_phases.zip)

gassmoeller commented:
This seems like a difference between your local machine and the official tester. Please have a look at the remainder of section 6.8.1 in the manual for reasons why tests sometimes differ slightly between machines. Or did you already create these results inside the official tester docker container? You can always upload your current version, in which case the official tester will run again and give you reliable test results.

Concerning your previous question: I agree that requiring to put in all parameters even if only some are needed seems unnecessary. Maybe you can modify the `parse_parameters` function to only read parameters that are needed? The most important thing for us is to keep existing parameter files working as before (backward compatibility) so if it is impossible to remove the duplication of input parameters without breaking old files than I would prefer to not change it.

lhy11009 commented:
@gassmoeller 
Hi Rene

a. "Please have a look at the remainder of section 6.8.1 in the manual for reasons "

Yes, I have read that section but I was not sure if that is the type of problem.


b. "Maybe you can modify the parse_parameters function to only read parameters that are needed?"

Thank's for your advice and I have modified the  parse_parameter function so that the duplication is avoided. 


Now, I think I will resubmit by tomorrow, let's look at it together then.
Have a nice day

lhy11009 commented:
@gassmoeller 

Hi Rene, I found it is not straightforward to rebase on the master branch since many of the phase transition functions have changed. I'll need more time to do this. I'll try to do it within this week and I may need some of your help.

Thanks

gassmoeller commented:
Yes there have been some changes in the material model in the last weeks. I will look over the PR next and leave you some comments. If you want we can meet virtually later this week or early next week and discuss how to move forward?

gassmoeller commented:
Concerning the rebase: You are right this is going to be complicated, because this PR was based on another branch that was modified and merged in the meantime. Here is how I would go about rebasing (after updating your master branch to the latest version):
- Create a backup copy of this branch (`git checkout -b backup`) then checkout this branch again. 
- Find the last common version between this branch and the master branch (`git merge-base master phase_transition_test`)
- Interactively rebase your branch to that commit (`git rebase -i REF` where REF is the result of the last command)
- During the interactive rebase remove all commits not made by you and squash all the commits made by you (https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history)
- Now rebase the branch to master. You will likely have to resolve a few conflicts, but at least you only have to resolve them once, and it should be only your changes that create conflicts (because mine were removed during the interactive rebase)

If you get stuck you can always delete the branch and restart from the backup branch.

gassmoeller commented on doc/modules/changes/20200305_li:
```diff
@@ -0,0 +1,16 @@
+New: Visco-plastic material model can now have phases in a single compostion leading to
```
```suggestion
New: The visco-plastic material model can now have multiple phases in a single composition leading to
```

gassmoeller commented on doc/modules/changes/20200305_li:
```diff
@@ -0,0 +1,16 @@
+Note that, for each composition, entries for properties should be one more than entries
+of phase transitions.
+If these entries are provided, the program will calculate density and viscosity by averaging
+among phases.
```
This is more detailed than what we usually put in our changelog. I would remove a lot and explain in the parameter documentation instead.

```suggestion
multiple phases and compositions in total.
If these entries are provided, the program will calculate density and viscosity by averaging
among the phases. Check the parameter documentation for the syntax to use multiple phases.
```

lhy11009 commented:
Sure, that would be most helpful. I'll like to first take a thorough look,
so maybe we can virtually meet Thursday or Friday? I reckon we'll need half
an hour. So would Thursday 2:00 - 2:30, 2:30 -3:00 or 3:00 -3:30 pm work
for you? Or anytime that fits your schedule, I am mostly available.

Stay healthy, have a nice day

On Mon, Mar 30, 2020 at 12:54 PM Rene Gassmoeller <notifications@github.com>
wrote:

> Yes there have been some changes in the material model in the last weeks.
> I will look over the PR next and leave you some comments. If you want we
> can meet virtually later this week or early next week and discuss how to
> move forward?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/geodynamics/aspect/pull/3313#issuecomment-606211018>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AG2WMFNOXOKFJX7NCFS24GDRKD2FZANCNFSM4JZBX53Q>
> .
>


-- 

Ph.D student of Geophysics
Earth & Planetary Sciences Dept., UC Davis
Davis, CA 95616

gassmoeller commented:
Great, I just sent you an email with a meeting link.

lhy11009 commented:
@gassmoeller 
I have rebased the branch, your strategy worked like a charm. I merged all commits into two. The first one is related to viscosity, second part is a more flexible input method to accept a single value in the subsection Material Model.

lhy11009 commented:
@gassmoeller 
The tests still failed, how can I deal with that?

lhy11009 commented:
@gassmoeller 
Also, what is the next step after I configured the docker and run the commit to give me:
dealii@709c3acef691
the folder doesn't seem to contain the regular aspect files as I am familiar with, how can I run a single test?

lhy11009 commented:
Thanks for reply

One issue is, if we split it, almost all the tests for visco_plastic_phases
will fail. So we have to recreate those tests if that is the case.

And, what is the next step when I get into that docker to run a single
test? I think I need first to run a single test in the same environment,
otherwise, it is hard to debug the tests

Have a nice day

On Fri, Apr 3, 2020 at 12:13 PM Rene Gassmoeller <notifications@github.com>
wrote:

> I have rebased the branch, your strategy worked like a charm. I merged all
> commits into two. The first one is related to viscosity, second part is a
> more flexible input method to accept a single value in the subsection
> Material Model.
>
> Ok, great. But I still see 3 commits (I think the first one should be part
> of the second one). You can fix that with another git rebase -i master.
> While at it, please open a separate PR for the last commit containing the
> flexible input method and remove it from this PR. This is not crucial
> functionality for this PR (just a convenience change), but it touches a
> very complicated function in utilities.cc and I would like to review that
> separately.
>
> The tests still failed, how can I deal with that?
>
> If you go to the test results (click on 'Details' next to the results,
> then go to the 'Tests' tab at the top) you see that 3 tests failed. 2
> existing tests and the one you added in this PR. For the one you added this
> is fine as long as you have verified on your system it is correct. That two
> existing tests failed means something changed for existing functionality
> and this is something you would need to investigate and fix. For now I
> would suggest to rebase first and split into two PRs then we know which of
> the changes broke the tests.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/geodynamics/aspect/pull/3313#issuecomment-608612269>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AG2WMFPAGLUH35HALDXNN5LRKYYNJANCNFSM4JZBX53Q>
> .
>


-- 

Ph.D student of Geophysics
Earth & Planetary Sciences Dept., UC Davis
Davis, CA 95616

lhy11009 commented:
@gassmoeller 
Actually, which is the test that is new there? Cause I hardly remember publishing any test my own. I would guess that is something that was originally there but got deleted later, does that make sense?
Sorry about the number of commits, I thought it was two -.-||

lhy11009 commented:
![4_3_2020](https://user-images.githubusercontent.com/28665365/78409589-e7b73100-75be-11ea-8ec1-b9156feaee71.PNG)

gassmoeller commented:
This is the new test: https://github.com/geodynamics/aspect/pull/3313/files#diff-6cfc973d2b573ad7e8f8026cfe2c3ec7

Inside the docker container you can do the following to run the test `visco_plastic_phases`:
```
cd aspect
mkdir tester-build
cd tester-build
cmake -G "Ninja" -D ASPECT_TEST_GENERATOR=Ninja -D ASPECT_USE_PETSC=OFF -D ASPECT_RUN_ALL_TESTS=ON -D ASPECT_PRECOMPILE_HEADERS=ON $SRC_PATH
ninja && ninja setup_tests
ctest -R visco_plastic_phases -V
```

And yes `vi` is not installed inside the container. It has only minimal software to make it small. You can use `cat` though.

lhy11009 commented:
@gassmoeller 
So when I execute the fourth command, it gives me the following error:
![4_3_2020_1](https://user-images.githubusercontent.com/28665365/78415019-f78d4000-75d3-11ea-8427-f5879ebd0b97.PNG)
Indeed this is no CMakeList.txt in the folder. Is that because there is something wrong with the docker?

gassmoeller commented:
yes that code was copy-pasted out of `contrib/utilities/compile_and_update_tests.sh` where SRC_PATH is previously set. Just use `cmake -G "Ninja" -D ASPECT_TEST_GENERATOR=Ninja -D ASPECT_USE_PETSC=OFF -D ASPECT_RUN_ALL_TESTS=ON -D ASPECT_PRECOMPILE_HEADERS=ON ..` instead, as you would do if you would run this outside of the docker container (ignoring the extra options). The last argument for `cmake` has to point to the aspect source directory that contains the CMakeLists.txt. If you already ran cmake successfully in this folder before, you can also point it to the directory where you are building, that is why `cmake .` also works in your existing build directories.

lhy11009 commented:
@gassmoeller 
Thanks, I figured this out. So one question is:

a. Do you usually maintain a separate aspect source for the docker test? 


And I want to ask your opinion on how to modify the test:

b. If we split this into two requests. Because they will fail the first request, Is it okay to rewrite those tests for visco_plastic_phases or we can do the two requests together and I can send you the code patches for the two of them to review?

c. the visco_plastic_derivatives tests suffer from error in small digits. Numbers in the screen-output file are within a 1e-4 relative error to the file for comparison. How should deal with it? Would I expect an error in code and try to look for it or Could I interpret this as a numerical issue and update the screen output as the new one?


Have a nice day

gassmoeller commented:
> a. Do you usually maintain a separate aspect source for the docker test?

No, I use one source directory, but separate build directories (on for a debug version, one for a release version, and one for the docker tester).

> c. the visco_plastic_derivatives tests suffer from error in small digits. Numbers in the screen-output file are within a 1e-4 relative error to the file for comparison. How should deal with it? Would I expect an error in code and try to look for it or Could I interpret this as a numerical issue and update the screen output as the new one?

We have a 1e-6 absolute error tolerance and a 1e-8 relative error tolerance for our tester comparison so your 1e-4 difference is counted as error. Also the `visco_plastic_derivatives` test compares analytical derivatives to numerical derivatives and does not depend on a full model run which is usually very accurate (most of the test happens inside the tests extra .cc file, not in the main aspect run, it is a bit complicated to read). Therefore it looks to me like something actually changed in the material model.

> b. If we split this into two requests. Because they will fail the first request, Is it okay to rewrite those tests for visco_plastic_phases or we can do the two requests together and I can send you the code patches for the two of them to review?

In this case we should handle the other PR first (the one that allows different input lengths), because afterwards the tests in this one will work immediately. Did I understand that correctly? Also if the last commit is what would go into that other PR I already see a few things I would ask you to generate separate PRs for (for example changing some parameters into vectors instead of scalars). The reasoning behind this is not lazyness on the part of the reviewer (me), but the fact that the code will be reviewed better, which means the code will be better afterwards (I found a nice article on the issue if you are interested in the reasoning behind it: https://www.thedroidsonroids.com/blog/splitting-pull-request).

lhy11009 commented:
@gassmoeller 

Sure, so first let me go past the tests. As long as I can pass all tests, there will be a way to deal with the pull request the way we like, which I leave there to be worried afterwards.

Have a nice day

lhy11009 commented:
@gassmoeller 
An update. To understand what went wrong in the test. I outputted all the compositional viscosities computed in the function calculate_isostrain_viscosity as well as the total viscosities computed by MaterialUtilities::average_value(). I put this in source codes in both the master branch and my branch and both to stderr instead of stdout. Then I run the test in separate build directories by directly running visco_plastic_derivatives.x.prm in the tests directory. At last, I compared the stderr output.

And guess what, surprisingly they are all the same! But the output in stdout is not the same. This is now one of the interesting bugs I have and I am still trying to understand what changed. After all, after I totally went through the test, I come to an understanding that the so-called 'analytical' approach did the same thing. They both called the 'evaluate' function and called it again after changing the pressure or strain to a small value. The difference is only that the 'analytical' approach does this in that compute_viscosity_derivatives function and the test approach does it explicitly out of it. They seem to be replicates of each other.

Moreover, in the process of this, I tried to understand the way the test and the               
compute_viscosity_derivatives function is assembled and get some questions. If it doesn't bother you too much, I would ask you. Otherwise, I'll find another person.

gassmoeller commented:
I have found the difference, or at least have a good clue where it comes from. Before the discussing the problem though I would like to make a few remarks:
1. This would be a good example to illustrate how to debug complicated bugs in aspect. Would you be ok if I use this as an example at one of the ASPECT meetings (possibly next week or in 3 weeks) to illustrate how to go about debugging? There were several small issues that made debugging complicated as you noticed.
2. This "bug" (it likely only affects the newton solver convergence) was largely caused by the fact that this PR mixed too many things to keep an overview of its main changes. Also in a proper review it very likely would have been found. So if you had separated the issues into different PRs and cleaned it up before trying to debug the test it would likely have saved you significant time (probably days?) and also saved me some time debugging (~2 hours). We all went into this trap at some point during our career, but it really pays off to keep changes as small and simple as possible (ask me about the particle code, or @MFraters about the newton solver at some point ;-)). Debugging 200 lines is much more complicated than debugging 100 lines twice, in particular if they mix different changes.

That being said, about the difference in the tests: `ViscoPlastic::evaluate` hands over `gamma_inputs` to the `calculate_isostrain_viscosities` function, but the other places in the material model that call  `calculate_isostrain_viscosities` do not do so (in particular the `compute_viscosity_derivatives` function). In theory that should not matter, because there are no phase transitions in this model, but the averaging inside `DislocationCreep::compute_viscosity` and `DiffusionCreep::compute_viscosity` is done differently if `gamma_inputs` is empty, from when it contains n entries that all say there is no phase transition. In other words we call `phase_average_value` even if there is not phase transition for the viscosity, but not for the derivatives. Now inside there nothing should happen, but if `by_log` is true we take the log and then invert the log, so maybe there we loose some accuracy, or there is some other bug if there is no phase transition. 

About solving this: We should either not call averaging if there is no phase transition, or make sure the averaging is always done (for viscosity and derivatives). How to best do this I would like to think about when we have a clear separation between the issue in this PR.

lhy11009 commented:
a. "In other words we call phase_average_value even if there is not phase
transition for the viscosity, but not for the derivatives. "

Yes, this seems to be a reasonable explanation. I am starting to see it.
Although I am still confused because I printed out the out.viscosities
which should be the total viscosity and that doesn't change either.

b. Would you be ok if I use this as an example at one of the ASPECT
meetings

Of course, it would certainly be interesting to look at an example. That
would be my first contribution to the community.

c. by the fact that this PR mixed too many things to keep an overview of
its main changes. Also in a proper review it very likely would have been
found. So if you had separated the issues into different PRs and cleaned it
up before trying to debug the test it would likely have saved you
significant time (probably days?) and also saved me some time debugging (~2
hours).

Yes, I agree. I'll remember to keep it small next time. I feel it helps
more to keep certain rules like that in mind than wondering what to do.

Thanks for your help, I know this would be extra work for you and I really
appreciate it. It's really a release for people to have a real supportive
resource when they are doing programming. Honestly for now, I'm not so sure
what I am doing when I am programming. However, I made a list of what to do
based on your suggestions. It's like a work flow and I'll be able to follow
that next time. 'Keep a PR small', I will certainly put it there. And I
would also put "Run the whole tests in docker after making change to each
section".

On Fri, Apr 17, 2020 at 12:40 PM Rene Gassmoeller <notifications@github.com>
wrote:

> I have found the difference, or at least have a good clue where it comes
> from. Before the discussing the problem though I would like to make a few
> remarks:
>
>    1. This would be a good example to illustrate how to debug complicated
>    bugs in aspect. Would you be ok if I use this as an example at one of the
>    ASPECT meetings (possibly next week or in 3 weeks) to illustrate how to go
>    about debugging? There were several small issues that made debugging
>    complicated as you noticed.
>    2. This "bug" (it likely only affects the newton solver convergence)
>    was largely caused by the fact that this PR mixed too many things to keep
>    an overview of its main changes. Also in a proper review it very likely
>    would have been found. So if you had separated the issues into different
>    PRs and cleaned it up before trying to debug the test it would likely have
>    saved you significant time (probably days?) and also saved me some time
>    debugging (~2 hours). We all went into this trap at some point during our
>    career, but it really pays off to keep changes as small and simple as
>    possible (ask me about the particle code, or @MFraters
>    <https://github.com/MFraters> about the newton solver at some point
>    ;-)). Debugging 200 lines is much more complicated than debugging 100 lines
>    twice, in particular if they mix different changes.
>
> That being said, about the difference in the tests: ViscoPlastic::evaluate
> hands over gamma_inputs to the calculate_isostrain_viscosities function,
> but the other places in the material model that call
> calculate_isostrain_viscosities do not do so (in particular the
> compute_viscosity_derivatives function). In theory that should not
> matter, because there are no phase transitions in this model, but the
> averaging inside DislocationCreep::compute_viscosity and
> DiffusionCreep::compute_viscosity is done differently if gamma_inputs is
> empty, from when it contains n entries that all say there is no phase
> transition. In other words we call phase_average_value even if there is
> not phase transition for the viscosity, but not for the derivatives. Now
> inside there nothing should happen, but if by_log is true we take the log
> and then invert the log, so maybe there we loose some accuracy, or there is
> some other bug if there is no phase transition.
>
> About solving this: We should either not call averaging if there is no
> phase transition, or make sure the averaging is always done (for viscosity
> and derivatives). How to best do this I would like to think about when we
> have a clear separation between the issue in this PR.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/geodynamics/aspect/pull/3313#issuecomment-615430170>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AG2WMFL3X3QKPWTKINCFZW3RNCWCVANCNFSM4JZBX53Q>
> .
>


-- 

Ph.D student of Geophysics
Earth & Planetary Sciences Dept., UC Davis
Davis, CA 95616

gassmoeller commented:
Awesome, here is how I would split up the branch:

1. Keep the current branch in the current state, you can use it as a starting point for the individual smaller PRs.

2. Make several new branches (one at a time) starting from your existing branch, and each having exactly one new functionality. Remove all the changes from these new branches that are not necessary for this one change. You can do that by looking through https://github.com/geodynamics/aspect/pull/3313/files and copying over the old state (the red lines) into your branch.

3. These are the branches and their functionalities I would make:

- add_phase_average_value function: This one would only have the new `phase_average_value` function, and the changes to `phase_average_equation_of_state_outputs`. This way you do not need new tests (our existing tests cover this function already).

- allow_single_values_in_value_array: This one would only have the changes to `parse_string_to_map`, and would include one test that uses the new functionality.

- use_phase_transition_in_diffusion_dislocation_creep: This one would depend on the two above and would need to wait until they are merged. It would include the changes to diffusion.cc, dislocation.cc, and visco_plastic.cc, and the test you already made.

lhy11009 commented:
Great, these are very clear.

I think what I will do is:

a. start from a clean mater branch, reimplement part of the total change(like a function or two). 

b. At the same time create tests. 

c. test in the docker to see it can pass all tests

d. add in comments

e. Do the last four steps iteratively

However, I have a thought that this process of first assembling things together regardless of the tests and neat features is also needed for myself's sake. Things like changing the parse function only show up in this process. I have a feeling that these procedures listed above should be practical for the next bundle of PRs I make.

I'll close this one and updates the small ones, perhaps in weeks, when I finish the prospectus I am writing right now.

Thanks, have a nice week
